### ISSUE TITLE
Set title based on resolved values

### ISSUE THREAD

**Issue by TheWrongAlice** [vinaygopinath/ngMeta#10]
In my product detail state I will resolve an API call to get a product, as seen below. Can I use this resolved `data` to set my `<title>` to be the product name?

http://pastebin.com/pxHVyh42

**Comment by vinaygopinath** [comment]
Untested, but this should work

``` javascript
      .state('product.details', {
        url : '/{id}',
        templateUrl : '/views/pages/product.html',
        controller: 'productCtrl',
        resolve: {
          data: function ($stateParams, soApi, ngMeta) {
            return soApi.getProduct({
              'id': $stateParams.id
            }).then(function(response) {
              ngMeta.setTitle(response.product.name);
              return response;
            });
          }
        }
      })
```

**Comment by TheWrongAlice** [comment]
Unfortunately no, because this code lives in the .config block of the app, so ngMeta is not available
(https://github.com/angular-ui/ui-router/wiki)

**Comment by vinaygopinath** [comment]
Ok, I figured it out with a bit of logging: ngMeta _can_ be injected into a resolve function, and `setTitle` is called properly, but right after the resolve function completes execution, ngMeta receives the `$stateChangeSuccess` broadcast and goes through the process of updating all tags, including the title, thus wiping out the title set in the resolve function.

As a temporary workaround, you can set the title from your `productCtrl`. I'm considering adding a `freezeTags` array property that can be optionally added to a state in cases involving resolve functions to keep some tags unaffected when the broadcast is received. Something like:

``` javascript
   resolve: {
     data: function($http, ...., ngMeta) {
       //Load data here
       ngMeta.setTitle('Some resolved value');
       ngMeta.setTag('description', 'Description of product - resolved from API call');
     }
   },
   meta: {
     freezeTags: ['title', 'description']
   }
```

What do you think?

**Comment by TheWrongAlice** [comment]
Ah, I understand.

Actually I think a nicer solution for me would be if I could just disable the listeners.
In configuration phase I could do
`ngMetaProvider.autoupdateOnStateChange(false); // Default: true`

Your init function would look like

```
var init = function() {
  $rootScope.ngMeta = {};
  if (config.autoupdateOnStateChange) {
    $rootScope.$on('$routeChangeSuccess', update);
    $rootScope.$on('$stateChangeSuccess', update);
  }
};
```

...so listeners would still be on as default, but I have the option of disabling them.
As a bonus, having fewer listeners will also slightly improve my site performance :)

If this sounds good to you, I can set up a PR :)

**Comment by vinaygopinath** [comment]
I like the idea, but by disabling the listeners, you would lose the ability to update meta tags in the `$routeProvider`/`$stateProvider` configuration for _all_ routes/states. You would have to programmatically set meta tags using `setTitle` and `setTag` through resolve functions, onEnter and controllers.

My intention with `freezeTags` was to allow `meta: { }` in state configurations while also preventing the `$stateChangeSuccess` listener from updating specific tags, so that in a case like this (based on the previous example)

```
   meta: {
     freezeTags: ['title', 'description'],
     image: 'http://example.com/image.png'
   }
```

The image tag would be set based on the state config, but title and description tags can be updated from resolve functions.

**Comment by vinaygopinath** [comment]
Perhaps the solution is to provide statewide update disabling (instead of app-wide update disabling) with `disableUpdate: true` and allowing users to set all meta tags programmatically for that particular state?

``` javascript
resolve: {
  data: function(..., ngMeta) {
    ....
    ngMeta.setTitle();
    ngMeta.setTag('description', '....');
    ngMeta.setTag('image', '....');
  }
},
meta: {
  disableUpdate: true
}
```

**Comment by TheWrongAlice** [comment]
Your last suggestion with `disableUpdate: true` would work for me. I am not sure how to go about implementing it in your code though. Where would I add the condition check?

**Comment by vinaygopinath** [comment]
I appreciate your PR offer - You can check for this condition in `readRouteMeta` and return the function right away if it's set.

**Comment by alexg313** [comment]
Thanks for including this feature. Definitely could use the disableUpdate feature. Any thoughts on when you'll push out a new release?

**Comment by vinaygopinath** [comment]
@alexg313 I just released [v0.3.8](https://github.com/vinaygopinath/ngMeta/releases/tag/v0.3.8) with `disableUpdate` support. Thanks everyone
